File: app/playRepository/BareCommit.java
/**
 * Yobi, Project Hosting SW
 *
 * Copyright 2014 NAVER Corp.
 * http://yobi.io
 *
 * @author Suwon Chae
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package playRepository;

import models.Project;
import models.User;
import org.eclipse.jgit.api.Git;
import org.eclipse.jgit.api.errors.ConcurrentRefUpdateException;
import org.eclipse.jgit.api.errors.JGitInternalException;
import org.eclipse.jgit.dircache.DirCache;
import org.eclipse.jgit.dircache.DirCacheBuilder;
import org.eclipse.jgit.dircache.DirCacheEditor;
import org.eclipse.jgit.dircache.DirCacheEntry;
import org.eclipse.jgit.internal.JGitText;
import org.eclipse.jgit.lib.*;
import org.eclipse.jgit.revwalk.RevCommit;
import org.eclipse.jgit.revwalk.RevWalk;
import org.eclipse.jgit.treewalk.CanonicalTreeParser;
import org.eclipse.jgit.treewalk.TreeWalk;
import utils.Config;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.channels.OverlappingFileLockException;
import java.text.MessageFormat;

import static org.eclipse.jgit.lib.Constants.HEAD;
import static org.eclipse.jgit.lib.Constants.OBJ_BLOB;
import static playRepository.BareRepository.findFileLineEnding;
import static utils.LineEnding.addEOL;
import static utils.LineEnding.changeLineEnding;

public class BareCommit {
    private PersonIdent personIdent;
    private Repository repository;
    private String commitMessage;
    private ObjectInserter objectInserter;

    private File file;
    private String refName;
    private ObjectId headObjectId;

    /**
     * Constructs a BareCommit.
     *
     * The given project MUST have a Git repository which has been CREATED.
     *
     * @param project  a project whose repository to be commited
     * @param user     the author and also the committer
     */
    public BareCommit(Project project, User user) {
        this.repository = BareRepository.getRepository(project);
        this.personIdent = new PersonIdent(user.name, user.email);
    }

    /**
     * Commit to bare repository with filename and text contents
     *
     * @param fileNameWithPath
     * @param contents
     * @param message
     */
    public ObjectId commitTextFile(
            String fileNameWithPath, String contents, String message) throws IOException {
        this.file = new File(fileNameWithPath);
        setCommitMessage(message);
        if (this.refName == null) {
            setRefName(HEAD);
        }

        ObjectId commitId = null;
        try {
            this.objectInserter = this.repository.newObjectInserter();
            contents = addEOL(changeLineEnding(contents, findFileLineEnding(repository, fileNameWithPath)));
            setHeadObjectId(this.refName);
            commitId = createCommitWithNewTree(createGitObjectWithText(contents));
            RefUpdate.Result result = refUpdate(commitId, refName);
        } catch (OverlappingFileLockException e) {
            play.Logger.error("Overlapping File Lock Error: " + e.getMessage());
        } finally {
            objectInserter.close();
            repository.close();
        }

        return commitId;
    }

    private boolean noHeadRef() {
        if(this.headObjectId == null) {
            return true;
        }
        return this.headObjectId.equals(ObjectId.zeroId());
    }

    private ObjectId createCommitWithNewTree(ObjectId targetTextFileObjectId) throws IOException {
        return objectInserter.insert(buildCommitWith(file.getName(), targetTextFileObjectId));
    }

    private CommitBuilder buildCommitWith(String fileName, ObjectId fileObjectId) throws IOException {
        CommitBuilder commit = new CommitBuilder();
        commit.setTreeId(createTreeWith(fileName, fileObjectId));
        if (!noHeadRef()) {
            commit.setParentId(this.headObjectId);
        }
        commit.setAuthor(getPersonIdent());
        commit.setCommitter(getPersonIdent());
        commit.setMessage(getCommitMessage());
        return commit;
    }

    private ObjectId createTreeWith(String fileName, ObjectId fileObjectId) throws IOException {
        if (noHeadRef()){
            return objectInserter.insert(newTreeWith(fileName, fileObjectId));
        } else {
            return objectInserter.insert(rebuildExistingTreeWith(fileName, fileObjectId));
        }
    }

    private TreeFormatter newTreeWith(String fileName, ObjectId fileObjectId) {
        TreeFormatter formatter = new TreeFormatter();
        formatter.append(fileName, FileMode.REGULAR_FILE, fileObjectId);
        return formatter;
    }

    private TreeFormatter rebuildExistingTreeWith(String fileName, ObjectId fileObjectId) throws IOException {
        TreeFormatter formatter = new TreeFormatter();
        CanonicalTreeParser treeParser = getCanonicalTreeParser(this.repository);

        boolean isInsertedInTree = false;
        while(!treeParser.eof()){
            String entryName = new String(treeParser.getEntryPathBuffer(), 0, treeParser.getEntryPathLength(), Config.getCharset());
            String nameForComparison = entryName;

            if (treeParser.getEntryFileMode() == FileMode.TREE){
                nameForComparison = entryName.concat("/"); //for tree ordering comparison
            }
            if (nameForComparison.compareTo(fileName) == 0 && isInsertedInTree == false){
                formatter.append(fileName, FileMode.REGULAR_FILE, fileObjectId);
                isInsertedInTree = true;
            } else if (nameForComparison.compareTo(fileName) > 0 && isInsertedInTree == false) {
                formatter.append(fileName, FileMode.REGULAR_FILE, fileObjectId);
                formatter.append(entryName.getBytes(Config.getCharset())
                        , treeParser.getEntryFileMode()
                        , treeParser.getEntryObjectId());
                isInsertedInTree = true;
            } else {
                formatter.append(entryName.getBytes(Config.getCharset())
                        , treeParser.getEntryFileMode()
                        , treeParser.getEntryObjectId());
            }

            treeParser = treeParser.next();
        }
        if(!isInsertedInTree){
            formatter.append(fileName, FileMode.REGULAR_FILE, fileObjectId);
        }
        return formatter;
    }

    private CanonicalTreeParser getCanonicalTreeParser(Repository repository) throws IOException {
        RevWalk revWalk = new RevWalk(repository);
        RevCommit commit = revWalk.parseCommit(this.headObjectId);
        return new CanonicalTreeParser(new byte[]{}, revWalk.getObjectReader(), commit.getTree().getId());
    }

    private ObjectId createGitObjectWithText(String contents) throws IOException {
        byte[] bytes = contents.getBytes(Config.getCharset());
        return objectInserter.insert(OBJ_BLOB, bytes, 0, bytes.length);
    }

    private RefUpdate.Result refUpdate(ObjectId commitId, String refName) throws IOException {
        RefUpdate ru = this.repository.updateRef(refName);
        ru.setForceUpdate(false);
        ru.setRefLogIdent(getPersonIdent());
        ru.setNewObjectId(commitId);
        if(hasOldCommit(refName)){
            ru.setExpectedOldObjectId(getCurrentMomentHeadObjectId());
        }
        ru.setRefLogMessage(getCommitMessage(), true);

        RefUpdate.Result result = ru.update();
        play.Logger.debug("Online commit: HEAD[" + this.headObjectId + "]:" + result + " New:" + commitId);
        return result;
    }

    private boolean hasOldCommit(String refName) throws IOException {
        return this.repository.findRef(refName).getObjectId() != null;
    }

    private PersonIdent getPersonIdent() {
        if (this.personIdent == null) {
            this.personIdent = new PersonIdent(this.repository);
        }
        return personIdent;
    }

    private String getCommitMessage() {
        if(this.commitMessage == null){
            return "Update " + this.file.getName();
        }
        return commitMessage;
    }

    public void setCommitMessage(String commitMessage) {
        this.commitMessage = commitMessage;
    }

    public void setHeadObjectId(String refName) throws IOException {
        if(this.repository.findRef(refName).getObjectId() == null){
            this.headObjectId = ObjectId.zeroId();
        } else {
            this.headObjectId = this.repository.findRef(refName).getObjectId();
        }
    }

    public ObjectId getCurrentMomentHeadObjectId() throws IOException {
        if( this.repository.findRef(refName).getObjectId() == null ){
            return ObjectId.zeroId();
        } else {
            return this.repository.findRef(refName).getObjectId();
        }
    }

    public void setRefName(String refName){
        this.refName = refName;
    }

    // Bare commit. It is referenced from https://gist.github.com/porcelli/3882505
    public ObjectId commitTextFile(final String branchName, final String path, String text, final String message) throws IOException {
        this.file = new File(this.repository.getDirectory(), path);
        org.apache.commons.io.FileUtils.writeStringToFile(this.file, text, "UTF-8");

        ObjectId commitId = null;
        Git git = new Git(this.repository);
        try (final ObjectInserter inserter = git.getRepository().newObjectInserter()) {
                // Create the in-memory index of the new/updated issue.
                this.headObjectId = git.getRepository().resolve(this.refName + "^{commit}");
                final DirCache index = createTemporaryIndex(git, this.headObjectId, path, file);
                final ObjectId indexTreeId = index.writeTree(inserter);

                // Create a commit object
                final CommitBuilder commit = getCommitBuilder(message, indexTreeId);

                // Insert the commit into the repository
                commitId = inserter.insert(commit);
                inserter.flush();

                final RefUpdate ru = getRefUpdate(branchName, commitId, git);
                final RefUpdate.Result rc = ru.forceUpdate();
                switch (rc) {
                    case NEW:
                    case FORCED:
                    case FAST_FORWARD:
                        break;
                    case REJECTED:
                    case LOCK_FAILURE:
                        throw new ConcurrentRefUpdateException(JGitText.get().couldNotLockHEAD, ru.getRef(), rc);
                    default:
                        throw new JGitInternalException(MessageFormat.format(JGitText.get().updatingRefFailed, Constants.HEAD, commitId.toString(), rc));
                }
        } catch (final Throwable t) {
            throw new RuntimeException(t);
        }

        return commitId;
    }

    private RefUpdate getRefUpdate(String branchName, ObjectId commitId, Git git) throws IOException {
        final RevWalk revWalk = new RevWalk(git.getRepository());
        final RevCommit revCommit = revWalk.parseCommit(commitId);
        final RefUpdate ru = git.getRepository().updateRef("refs/heads/" + branchName);
        if (this.headObjectId == null) {
            ru.setExpectedOldObjectId(ObjectId.zeroId());
        } else {
            ru.setExpectedOldObjectId(this.headObjectId);
        }
        ru.setNewObjectId(commitId);
        ru.setRefLogMessage("commit: " + revCommit.getShortMessage(), false);
        revWalk.close();
        return ru;
    }

    private CommitBuilder getCommitBuilder(String message, ObjectId indexTreeId) {
        final CommitBuilder commit = new CommitBuilder();
        commit.setAuthor(this.getPersonIdent());
        commit.setCommitter(this.getPersonIdent());
        commit.setEncoding(Constants.CHARACTER_ENCODING);
        commit.setMessage(message);
        //headId can be null if the repository has no commit yet
        if (this.headObjectId != null) {
            commit.setParentId(this.headObjectId);
        }
        commit.setTreeId(indexTreeId);
        return commit;
    }

    private static DirCache createTemporaryIndex(final Git git, final ObjectId headId, final String path, final File file) {
        final DirCache inCoreIndex = DirCache.newInCore();
        final DirCacheBuilder dcBuilder = inCoreIndex.builder();
        final ObjectInserter inserter = git.getRepository().newObjectInserter();

        try {
            if (file != null) {
                final DirCacheEntry dcEntry = new DirCacheEntry(path);
                dcEntry.setLength(file.length());
                dcEntry.setLastModified(file.lastModified());
                dcEntry.setFileMode(FileMode.REGULAR_FILE);

                final InputStream inputStream = new FileInputStream(file);
                try {
                    dcEntry.setObjectId(inserter.insert(Constants.OBJ_BLOB, file.length(), inputStream));
                } finally {
                    inputStream.close();
                }

                dcBuilder.add(dcEntry);
            }

            if (headId != null) {
                final TreeWalk treeWalk = new TreeWalk(git.getRepository());
                final int hIdx = treeWalk.addTree(new RevWalk(git.getRepository()).parseTree(headId));
                treeWalk.setRecursive(true);

                while (treeWalk.next()) {
                    final String walkPath = treeWalk.getPathString();
                    final CanonicalTreeParser hTree = treeWalk.getTree(hIdx, CanonicalTreeParser.class);

                    if (!walkPath.equals(path)) {
                        // add entries from HEAD for all other paths
                        // create a new DirCacheEntry with data retrieved from HEAD
                        final DirCacheEntry dcEntry = new DirCacheEntry(walkPath);
                        dcEntry.setObjectId(hTree.getEntryObjectId());
                        dcEntry.setFileMode(hTree.getEntryFileMode());

                        // add to temporary in-core index
                        dcBuilder.add(dcEntry);
                    }
                }
                treeWalk.close();
            }

            dcBuilder.finish();
        } catch (Exception e) {
            throw new RuntimeException(e);
        } finally {
            inserter.close();
        }

        if (file == null) {
            final DirCacheEditor editor = inCoreIndex.editor();
            editor.add(new DirCacheEditor.DeleteTree(path));
            editor.finish();
        }

        return inCoreIndex;
    }
}

File: src/main/kotlin/com/github/yonaprojects/yona/domain/vcs/BareCommit.kt
package com.github.yonaprojects.yona.domain.vcs

import com.github.yonaprojects.yona.domain.project.Project
import com.github.yonaprojects.yona.domain.user.User
import org.eclipse.jgit.api.Git
import org.eclipse.jgit.api.errors.ConcurrentRefUpdateException
import org.eclipse.jgit.api.errors.JGitInternalException
import org.eclipse.jgit.dircache.DirCache
import org.eclipse.jgit.dircache.DirCacheEntry
import org.eclipse.jgit.internal.JGitText
import org.eclipse.jgit.lib.*
import org.eclipse.jgit.revwalk.RevCommit
import org.eclipse.jgit.revwalk.RevWalk
import org.eclipse.jgit.storage.file.FileRepositoryBuilder
import org.eclipse.jgit.treewalk.CanonicalTreeParser
import org.eclipse.jgit.treewalk.TreeWalk
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.text.MessageFormat
import java.time.Instant

class BareCommit(
    project: Project,
    user: User,
    gitBaseDir: String,
    defaultBranch: String = "main",
    // 위키 저장소(`<owner>/<project>.wiki.git`)처럼 프로젝트의 실제 코드 저장소가 아닌 다른
    // bare 저장소에 커밋해야 하는 호출부를 위한 오버라이드. null이면 기존과 동일하게
    // "${project.name}.git"을 그대로 쓴다(기존 호출부 전부 무변경).
    repoNameOverride: String? = null
) {
    private val repository: Repository
    private val personIdent: PersonIdent
    private var commitMessage: String? = null
    private var file: File? = null
    // 새 프로젝트의 실제 초기 브랜치(GitRepository.create()가 만드는 것)와 일치해야 한다 — 안
    // 그러면 setRefName() 없이 이 기본값으로 커밋하는 경로(README/ISSUE_TEMPLATE 등)가 실제 HEAD가
    // 가리키는 빈 브랜치가 아니라 별도의 고아 브랜치를 만들어버린다.
    private var refName: String = Constants.R_HEADS + defaultBranch
    private var headObjectId: ObjectId? = null

    init {
        val repoName = repoNameOverride ?: project.name
        val gitDir = File(File(gitBaseDir), "${project.owner}/$repoName.git")
        this.repository = FileRepositoryBuilder().setGitDir(gitDir).build()
        this.personIdent = PersonIdent(user.name ?: user.loginId, user.email ?: "")
    }

    fun setRefName(refName: String) {
        this.refName = refName
    }

    @Throws(IOException::class)
    fun commitTextFile(fileNameWithPath: String, contents: String, message: String): ObjectId? {
        this.file = File(fileNameWithPath)
        this.commitMessage = message
        
        var commitId: ObjectId? = null
        val inserter = repository.newObjectInserter()
        try {
            val ref = repository.findRef(refName)
            this.headObjectId = ref?.objectId ?: ObjectId.zeroId()
            
            val blobId = inserter.insert(Constants.OBJ_BLOB, contents.toByteArray(Charsets.UTF_8))
            val treeId = createTreeWith(inserter, file!!.name, blobId)
            
            val commitBuilder = CommitBuilder().apply {
                setTreeId(treeId)
                if (headObjectId != ObjectId.zeroId()) {
                    setParentId(headObjectId)
                }
                author = personIdent
                committer = personIdent
                setMessage(commitMessage)
            }
            
            commitId = inserter.insert(commitBuilder)
            inserter.flush()
            
            val ru = repository.updateRef(refName)
            ru.isForceUpdate = false
            ru.setRefLogIdent(personIdent)
            ru.setNewObjectId(commitId)
            if (ref?.objectId != null) {
                ru.setExpectedOldObjectId(headObjectId)
            }
            ru.setRefLogMessage(commitMessage, true)
            ru.update()
        } finally {
            inserter.close()
            repository.close()
        }
        return commitId
    }

    // yona BareCommit.java의 bare commit(https://gist.github.com/porcelli/3882505 인용) 대응.
    // 위 3-인자 commitTextFile()과 달리 1) branchName으로 지정한 refs/heads/<branchName>에만
    // 커밋을 반영하고, 2) DirCache+재귀 TreeWalk로 기존 트리의 모든 하위 경로(nested path) 파일을 보존한 채
    // path(중첩 경로 가능) 위치에 새 파일을 반영한다. GitUtil.commitTextFile()가 setRefName(branchName)을
    // 먼저 호출한 뒤 이 메서드를 호출하는 것을 전제로 한다(yona와 동일한 사용 계약).
    @Throws(IOException::class)
    fun commitTextFile(branchName: String, path: String, text: String, message: String): ObjectId? {
        this.file = File(repository.directory, path)
        this.file!!.parentFile?.mkdirs()
        this.file!!.writeText(text, Charsets.UTF_8)

        var commitId: ObjectId? = null
        val git = Git(repository)
        try {
            git.repository.newObjectInserter().use { inserter ->
                this.headObjectId = git.repository.resolve("$refName^{commit}")
                val index = createTemporaryIndex(git, headObjectId, path, file!!)
                val indexTreeId = index.writeTree(inserter)

                val commit = getCommitBuilder(message, indexTreeId)

                commitId = inserter.insert(commit)
                inserter.flush()

                val ru = getRefUpdate(branchName, commitId!!, git)
                when (val rc = ru.forceUpdate()) {
                    RefUpdate.Result.NEW, RefUpdate.Result.FORCED, RefUpdate.Result.FAST_FORWARD -> {}
                    RefUpdate.Result.REJECTED, RefUpdate.Result.LOCK_FAILURE ->
                        throw ConcurrentRefUpdateException(JGitText.get().couldNotLockHEAD, ru.ref, rc)
                    else ->
                        throw JGitInternalException(
                            MessageFormat.format(JGitText.get().updatingRefFailed, Constants.HEAD, commitId.toString(), rc)
                        )
                }
            }
        } catch (t: Throwable) {
            throw RuntimeException(t)
        }

        return commitId
    }

    // 위키 페이지 저장(생성/수정/이름변경 통합) 대응. Forgejo 위키 편집 폼처럼 제목(=경로)과
    // 본문을 한 커밋으로 함께 반영한다. oldPath가 null이면 신규 생성, oldPath != newPath면 그
    // 경로의 파일을 지우고 newPath에 새로 쓰는 이름변경까지 한 커밋에서 처리한다. 위 3-인자
    // commitTextFile()과 달리 실제 파일시스템(bare 저장소 디렉터리)에 스크래치 파일을 쓰지 않고
    // in-core DirCache만으로 처리한다 — 위키 저장소는 코드브라우저 온라인편집과 달리 동시에
    // 여러 페이지가 저장될 수 있어, 디스크에 임시 파일을 남기는 기존 방식보다 안전하다.
    @Throws(IOException::class)
    fun commitPage(branchName: String, oldPath: String?, newPath: String, text: String, message: String): ObjectId? {
        var commitId: ObjectId? = null
        val git = Git(repository)
        try {
            git.repository.newObjectInserter().use { inserter ->
                this.headObjectId = git.repository.resolve("$refName^{commit}")
                val index = createPageIndex(git, headObjectId, oldPath, newPath, text, inserter)
                val indexTreeId = index.writeTree(inserter)

                val commit = getCommitBuilder(message, indexTreeId)

                commitId = inserter.insert(commit)
                inserter.flush()

                val ru = getRefUpdate(branchName, commitId!!, git)
                applyRefUpdate(ru, commitId!!)
            }
        } catch (t: Throwable) {
            throw RuntimeException(t)
        }
        return commitId
    }

    // 위키 페이지 삭제 대응. path에 해당하는 트리 엔트리만 제거한 새 커밋을 만든다.
    @Throws(IOException::class)
    fun deletePage(branchName: String, path: String, message: String): ObjectId? {
        var commitId: ObjectId? = null
        val git = Git(repository)
        try {
            git.repository.newObjectInserter().use { inserter ->
                this.headObjectId = git.repository.resolve("$refName^{commit}")
                    ?: throw java.io.FileNotFoundException("위키 페이지를 찾을 수 없습니다(빈 저장소): $path")
                val index = createPageIndex(git, headObjectId, oldPath = path, newPath = "", text = null, inserter = inserter)
                val indexTreeId = index.writeTree(inserter)

                val commit = getCommitBuilder(message, indexTreeId)

                commitId = inserter.insert(commit)
                inserter.flush()

                val ru = getRefUpdate(branchName, commitId!!, git)
                applyRefUpdate(ru, commitId!!)
            }
        } catch (t: Throwable) {
            throw RuntimeException(t)
        }
        return commitId
    }

    private fun applyRefUpdate(ru: RefUpdate, commitId: ObjectId) {
        when (val rc = ru.forceUpdate()) {
            RefUpdate.Result.NEW, RefUpdate.Result.FORCED, RefUpdate.Result.FAST_FORWARD -> {}
            RefUpdate.Result.REJECTED, RefUpdate.Result.LOCK_FAILURE ->
                throw ConcurrentRefUpdateException(JGitText.get().couldNotLockHEAD, ru.ref, rc)
            else ->
                throw JGitInternalException(
                    MessageFormat.format(JGitText.get().updatingRefFailed, Constants.HEAD, commitId.toString(), rc)
                )
        }
    }

    // HEAD 트리를 in-core DirCache로 복사하면서 oldPath(있으면)/newPath 두 경로를 제외한 뒤,
    // text가 null이 아니면 newPath에 새 blob을 추가한다(text가 null이면 삭제 전용 — newPath는
    // 빈 문자열을 넘겨 아무것도 추가하지 않는다, deletePage() 참고).
    private fun createPageIndex(
        git: Git,
        headId: ObjectId?,
        oldPath: String?,
        newPath: String,
        text: String?,
        inserter: ObjectInserter
    ): DirCache {
        val inCoreIndex = DirCache.newInCore()
        val dcBuilder = inCoreIndex.builder()

        if (headId != null) {
            val revWalk = RevWalk(git.repository)
            val treeWalk = TreeWalk(git.repository)
            val hIdx = treeWalk.addTree(revWalk.parseTree(headId))
            treeWalk.isRecursive = true
            while (treeWalk.next()) {
                val walkPath = treeWalk.pathString
                if (walkPath == oldPath || walkPath == newPath) continue
                val hTree = treeWalk.getTree(hIdx, CanonicalTreeParser::class.java)
                val dcEntry = DirCacheEntry(walkPath)
                dcEntry.setObjectId(hTree.entryObjectId)
                dcEntry.fileMode = hTree.entryFileMode
                dcBuilder.add(dcEntry)
            }
            treeWalk.close()
        }

        if (text != null) {
            val blobId = inserter.insert(Constants.OBJ_BLOB, text.toByteArray(Charsets.UTF_8))
            val newEntry = DirCacheEntry(newPath)
            newEntry.setObjectId(blobId)
            newEntry.fileMode = FileMode.REGULAR_FILE
            dcBuilder.add(newEntry)
        }

        dcBuilder.finish()
        return inCoreIndex
    }

    private fun getRefUpdate(branchName: String, commitId: ObjectId, git: Git): RefUpdate {
        val revWalk = RevWalk(git.repository)
        val revCommit = revWalk.parseCommit(commitId)
        val ru = git.repository.updateRef(Constants.R_HEADS + branchName)
        if (this.headObjectId == null) {
            ru.setExpectedOldObjectId(ObjectId.zeroId())
        } else {
            ru.setExpectedOldObjectId(this.headObjectId)
        }
        ru.setNewObjectId(commitId)
        ru.setRefLogMessage("commit: " + revCommit.shortMessage, false)
        revWalk.close()
        return ru
    }

    private fun getCommitBuilder(message: String, indexTreeId: ObjectId): CommitBuilder {
        val commit = CommitBuilder()
        commit.author = this.personIdent
        commit.committer = this.personIdent
        commit.encoding = Charsets.UTF_8
        commit.message = message
        // headObjectId는 저장소에 커밋이 아직 없는 경우 null일 수 있다
        if (this.headObjectId != null) {
            commit.setParentId(this.headObjectId)
        }
        commit.setTreeId(indexTreeId)
        return commit
    }

    private fun createTemporaryIndex(git: Git, headId: ObjectId?, path: String, file: File): DirCache {
        val inCoreIndex = DirCache.newInCore()
        val dcBuilder = inCoreIndex.builder()
        val inserter = git.repository.newObjectInserter()

        try {
            val dcEntry = DirCacheEntry(path)
            dcEntry.setLength(file.length())
            dcEntry.setLastModified(Instant.ofEpochMilli(file.lastModified()))
            dcEntry.fileMode = FileMode.REGULAR_FILE

            FileInputStream(file).use { inputStream ->
                dcEntry.setObjectId(inserter.insert(Constants.OBJ_BLOB, file.length(), inputStream))
            }

            dcBuilder.add(dcEntry)

            if (headId != null) {
                val treeWalk = TreeWalk(git.repository)
                val hIdx = treeWalk.addTree(RevWalk(git.repository).parseTree(headId))
                treeWalk.isRecursive = true

                while (treeWalk.next()) {
                    val walkPath = treeWalk.pathString
                    val hTree = treeWalk.getTree(hIdx, CanonicalTreeParser::class.java)

                    if (walkPath != path) {
                        // HEAD의 다른 모든 경로를 임시 in-core 인덱스에 그대로 옮겨온다
                        val dcEntry2 = DirCacheEntry(walkPath)
                        dcEntry2.setObjectId(hTree.entryObjectId)
                        dcEntry2.fileMode = hTree.entryFileMode
                        dcBuilder.add(dcEntry2)
                    }
                }
                treeWalk.close()
            }

            dcBuilder.finish()
        } finally {
            inserter.close()
        }

        return inCoreIndex
    }

    private fun createTreeWith(inserter: ObjectInserter, fileName: String, fileObjectId: ObjectId): ObjectId {
        return if (headObjectId == ObjectId.zeroId() || headObjectId == null) {
            val formatter = TreeFormatter().apply {
                append(fileName, FileMode.REGULAR_FILE, fileObjectId)
            }
            inserter.insert(formatter)
        } else {
            val formatter = TreeFormatter()
            val revWalk = RevWalk(repository)
            val commit = revWalk.parseCommit(headObjectId)
            val treeParser = CanonicalTreeParser(byteArrayOf(), revWalk.objectReader, commit.tree.id)
            
            var isInserted = false
            while (!treeParser.eof()) {
                val entryName = String(treeParser.entryPathBuffer, 0, treeParser.entryPathLength, Charsets.UTF_8)
                var nameForComparison = entryName
                if (treeParser.entryFileMode == FileMode.TREE) {
                    nameForComparison = "$entryName/"
                }
                
                if (nameForComparison.compareTo(fileName) == 0 && !isInserted) {
                    formatter.append(fileName, FileMode.REGULAR_FILE, fileObjectId)
                    isInserted = true
                } else if (nameForComparison.compareTo(fileName) > 0 && !isInserted) {
                    formatter.append(fileName, FileMode.REGULAR_FILE, fileObjectId)
                    formatter.append(entryName.toByteArray(Charsets.UTF_8), treeParser.entryFileMode, treeParser.entryObjectId)
                    isInserted = true
                } else {
                    formatter.append(entryName.toByteArray(Charsets.UTF_8), treeParser.entryFileMode, treeParser.entryObjectId)
                }
                treeParser.next()
            }
            if (!isInserted) {
                formatter.append(fileName, FileMode.REGULAR_FILE, fileObjectId)
            }
            inserter.insert(formatter)
        }
    }
}
